pkg/fleetautoscalers/fleetautoscalers.go

```diff
@@ -224,3 +234,425 @@ func applyBufferPolicy(b *autoscalingv1.BufferPolicy, f *agonesv1.Fleet) (int32,
+			return scaleUpByInteger(replicas, capacity, count,
+				aggCapacity, availableCapacity, maxCapacity, buffer)
+		case availableCapacity > buffer: // Scale Down
+			if limited && scale == 1 { // Case where we want to scale down but we're already limited by MinCapacity
+				return scaleLimited(scale, f, gameServerLister, nodeCounts, key, isCounter, replicas,
+					capacity, aggCapacity, minCapacity, maxCapacity)
+			}
+			return scaleDownByInteger(f, gameServerLister, nodeCounts, key, isCounter, replicas,
+				availableCapacity, aggCount, aggCapacity, minCapacity, buffer)
+		}
+	}
+
+	// Desired replicas based on BufferSize specified as a percent (i.e. 5%)
+	bufferPercent, err := intstr.GetValueFromIntOrPercent(&bufferSize, 100, isCounter)
+	if err != nil {
+		return 0, false, err
+	}
+	// The desired total capacity across the fleet (see applyBufferPolicy for explanation)
+	desiredCapacity := float64(aggCount*100) / float64(100-bufferPercent)
+
+	switch roundedDesiredCapacity := int64(math.Ceil(desiredCapacity)); {
+	case roundedDesiredCapacity == aggCapacity:
+		if limited {
+			return scaleLimited(scale, f, gameServerLister, nodeCounts, key, isCounter, replicas,
+				capacity, aggCapacity, minCapacity, maxCapacity)
+		}
+		return replicas, false, nil
+
+	case roundedDesiredCapacity > aggCapacity: // Scale up
+		if limited && scale == -1 { // Case where we want to scale up but we're already limited by MaxCapacity
+			return scaleLimited(scale, f, gameServerLister, nodeCounts, key, isCounter, replicas,
+				capacity, aggCapacity, minCapacity, maxCapacity)
+		}
+		return scaleUpByPercent(replicas, count, aggCount, capacity, aggCapacity, maxCapacity,
+			desiredCapacity, float64(bufferPercent))
+
+	case roundedDesiredCapacity < aggCapacity: // Scale down
+		if limited && scale == 1 { // Case where we want to scale down but we're already limited by MinCapacity
+			return scaleLimited(scale, f, gameServerLister, nodeCounts, key, isCounter, replicas,
+				capacity, aggCapacity, minCapacity, maxCapacity)
+		}
+		return scaleDownByPercent(f, gameServerLister, nodeCounts, key, isCounter, replicas,
+			aggCount, aggCapacity, minCapacity, float64(bufferPercent))
+	}
+
+	if isCounter {
+		return 0, false, errors.Errorf("unable to apply CounterPolicy %v", c)
+	}
+	return 0, false, errors.Errorf("unable to apply ListPolicy %v", l)
+}
+
+// getSortedGameServers returns the list of Game Servers for the Fleet in the order in which the
+// Game Servers would be deleted.
+func getSortedGameServers(f *agonesv1.Fleet, gameServerLister listeragonesv1.GameServerLister,
+	nodeCounts map[string]gameservers.NodeCount) ([]*agonesv1.GameServer, error) {
+	// TODO: Should we handle this differently for strategy Distributed?
+	gss := f.GetGameServerSpec()
+	gsList, err := fleets.ListGameServersByFleetOwner(gameServerLister, f)
+	if err != nil {
+		return nil, err
+	}
+	gameServers := gsSets.SortGameServersByStrategy(gss.Scheduling, gsList, nodeCounts, f.Spec.Priorities)
```

Comment:
Does the Fleet have the scheduling - do you need the GameServerSet here?

I think we chatted about this - should `SortGameServersByStrategy` move into some kind of common place? (Maybe we need a `/pkg/sorting`  or something like that?) Currently packages that are controllers don't tend to bleed into each other. I'm wondering if we want to maintain that seperation.

Reply:
I tried to move the SortGameServersByStrategy to common.go but could not because of cyclical dependencies. Is it best to put it in a separate package? Would we still need the gameServerLister and nodeCounter passed into the Fleetautoscaler, or would that be kept entirely separate?

Reply:
> I tried to move the SortGameServersByStrategy to common.go but could not because of cyclical dependencies.

Yeah, that makes sense.

> Is it best to put it in a separate package?

That's sorta what I'm thinking - move both `SortGameServersByStrategy` and probably the `PerNodeCounter` into there (since they are so tied together), and we can have a central place for sorting strategies. Otherwise we seem to be arbitrarily importing from `pkg/gameservers` and `pkg/gameserversets` which feels a little messy to me.

(This refactor could be a separate PR against the current code base if you wanted to split the work out, and carve some chunks off this one).

Reply:
OK, I will make changes to sorting by Priorities for Strategy distributed.